feat: add support for FGAC spanner

hemanshv opened the pull request:
Hi @jskeet and @amanda-tarafa ,

This is a prototype PR for Spanner FGAC  with the approach suggested by Amanda. 
Its still work in progress. Please have a look and provide your valuable feedback.
Thanks,

amanda-tarafa commented:
I'll take a look when I'm back, or maybe before if I can.

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.TargetedSessionPool.cs:
```diff
@@ -117,7 +117,10 @@ internal TargetedSessionPool(SessionPool parent, DatabaseName databaseName, bool
                     }
                 };
-
+                if (databaseRole != null)
```
I'd just use `CreatorRole = databaseRole ?? ""` in the object initializer above.

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.TargetedSessionPool.cs:
```diff
@@ -858,7 +861,8 @@ internal DatabaseStatistics GetStatisticsSnapshot()
                         Healthy,
-                        Shutdown);
+                        Shutdown,
+                        _batchCreateSessionRequestTemplate.SessionTemplate.CreatorRole);
```
I'd move this to just below `_databaseName` (assuming it's still a separate parameter). That's what it logically belong with.

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SessionPoolManager.cs:
```diff
@@ -161,12 +161,12 @@ internal void Release(SessionPool sessionPool)
             _targetedPools.ToArray().Select(tp => tp.Value.GetStatisticsSnapshot()).ToList().AsReadOnly();
 
-        internal SessionPool.DatabaseStatistics GetDatabaseStatistics(SpannerClientCreationOptions options, DatabaseName databaseName)
+        internal SessionPool.DatabaseStatistics GetDatabaseStatistics(SpannerClientCreationOptions options, DatabaseName databaseName, string databaseRole)
```
The fact that all these methods are changing in the same way strongly suggests to me that we want a new type to represent "the key for a targeted pool within a SessionPool".

The naming of that could be interesting - we have TargetedSessionPool at the moment, but that's internal.

I suggest you change this PR to create such a type (currently with just a database name and creator role) and we can worry about the naming afterwards.

jskeet replied:
I now see it already exists as SessionPoolInfo - I'll add more comments there, but basically I think it should be made public and exposed in the APIs.

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.DatabaseStatistics.cs:
```diff
@@ -91,9 +96,11 @@ public sealed class DatabaseStatistics
                 bool healthy,
-                bool shutdown)
+                bool shutdown,
+                string databaseRole)
```
With the new pool key class, I'd expect the constructor to take that, and store it in a field - the DatabaseName and DatabaseRole properties can then just delegate to that (e.g. `public string DatabaseName => key.DatabaseName;`). We probably don't need to expose the key as a public property here. (It does need to be in the public API so that the Data package can use it, but I don't expect users to need it much.)

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.cs:
```diff
@@ -36,6 +36,44 @@ namespace Google.Cloud.Spanner.V1
+        /// This struct contains information required to identify the database sessions.
+        /// It will be used as key in <see cref="_targetedPools"/>.
+        /// </summary>
+        internal readonly struct SessionPoolInfo
```
As noted earlier, I'd make this public. But I'd rename it to something involving "key" (and it's not just SessionPoolKey, as it's the key of a pool *within* a session pool). It's not really information about the session pool itself.

jskeet replied:
This should implement `IEquatable<T>` for itself as well, to avoid boxing.

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.cs:
```diff
@@ -36,6 +36,44 @@ namespace Google.Cloud.Spanner.V1
+            internal DatabaseName DatabaseName { get; }
+
+            /// <summary>
+            /// Database role of the associated <see cref="TargetedSessionPool"/>.
```
I think we may find it simplest to guarantee that this isn't null - if we make it an empty string everywhere that there's no database role, it'll make all kinds of operations a bit simpler. I think it's reasonable to do that in the constructor, just with `DatabaseRole = databaseRole ?? ""`.

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.cs:
```diff
@@ -36,6 +36,44 @@ namespace Google.Cloud.Spanner.V1
+                {
+                    return false;
+                }
+                return obj.ToString() == ToString();
```
We shouldn't perform string formatting every time. See https://github.com/googleapis/gax-dotnet/blob/main/Google.Api.Gax/GaxEqualityHelpers.cs for helpers for this.

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.cs:
```diff
@@ -36,6 +36,44 @@ namespace Google.Cloud.Spanner.V1
+            /// </summary>
+            internal string DatabaseRole { get; }
+
+            internal SessionPoolInfo(DatabaseName databaseName, string dataabaseRole)
```
Nit: dataabase => database

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.cs:
```diff
@@ -36,6 +36,44 @@ namespace Google.Cloud.Spanner.V1
+            }
+
+            /// <inheritdoc/>
+            public override string ToString() => $"Database Name = {DatabaseName};Database role = {DatabaseRole}";
```
Nit: I'd add a space after the semi-colon in the string, and probably change the "key" parts to the property names, so `$"DatabaseName = {DatabaseName}; DatabaseRole = {DatabaseRole}"`

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.cs:
```diff
@@ -135,10 +173,17 @@ private void MaintainPool()
+        /// Provides a snapshot of statistics for a database-databaseRole-specific pool.
+        /// </summary>
+        /// <returns>A snapshot of statistics for this pool.</returns>
+        public DatabaseStatistics GetStatisticsSnapshot(DatabaseName databaseName, string databaseRole)
```
Let's make the new overload just accept the new key type, and create a new key in the method above.

jskeet replied:
(Ditto for the other methods.)

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.cs:
```diff
@@ -205,10 +262,24 @@ public PooledSession CreateDetachedSession(SessionName sessionName, ByteString t
+            WhenPoolReady(databaseName, null, cancellationToken);
+
+        /// <summary>
+        /// Returns a task indicating when the session pool associated with the given database name is populated up to its minimum size.
```
We'll need to update the comment here - presumably to "with the given key" or similar. If we keep this precise signature, then the comment should be "with the given database name and role".

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.cs:
```diff
@@ -223,12 +294,27 @@ public Task WhenPoolReady(DatabaseName databaseName, CancellationToken cancellat
+            ShutdownPoolAsync(databaseName, null, cancellationToken);
+
+        /// <summary>
+        /// Shuts down the session pool associated with the given database name.
```
Ditto in terms of comments.

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.cs:
```diff
@@ -36,6 +36,59 @@ namespace Google.Cloud.Spanner.V1
+        /// This struct contains information required to identify the SessionPool.
+        /// It will be used as key in <see cref="_targetedPools"/>.
+        /// </summary>
+        public readonly struct PartitionKey: IEquatable<PartitionKey>
```
I'd suggest moving this into its own file - SessionPool.PartitionKey.cs (to match the other nested types).

Also, please add a space before the colon.

hemanshv replied:
Ok Jon will update

jskeet commented:
To what extent is this ready for an actual full review? I'd rather not do several rounds of review when it's not expected to be ready to merge.

hemanshv commented:
Hey @jskeet  @amanda-tarafa,
I have updated the PR with the latest changes for FGAC.
I saw that there are some tests are failing. I am looking into this.
I will let you know once the tests are fixed.

Thanks,

amanda-tarafa commented:
@hemanshv Please also make sure that you have run all integration tests locally and that they are all green before sending the PR for review. Thanks!

hemanshv commented:
Hey @jskeet @amanda-tarafa 
The issue with the Integration tests are now fixed.
Please review the PR and provide your valuable feedback.
Thanks,

amanda-tarafa commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.DatabaseStatistics.cs:
```diff
@@ -29,6 +29,16 @@ public sealed class DatabaseStatistics
+           /// <summary>
+           /// Partition Key of sessions this set of statistics corresponds to.
+           /// </summary>
+            private PartitionKey partitionKey { get; }
```
If this is a property, name with PascalCase event if it's private. If it's an internal variable, name with _camelCase.

amanda-tarafa commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.DatabaseStatistics.cs:
```diff
@@ -29,6 +29,16 @@ public sealed class DatabaseStatistics
             /// </summary>
             public DatabaseName DatabaseName { get; }
```
```suggestion
            public DatabaseName DatabaseName { get; } => PartitionKey.DatabaseName;
```

amanda-tarafa commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.DatabaseStatistics.cs:
```diff
@@ -29,6 +29,16 @@ public sealed class DatabaseStatistics
+            /// <summary>
+            /// Database Role of sessions this set of statistics corresponds to. Can be null.
+            /// </summary>
+            public string DatabaseRole { get; }
```
```suggestion
            public string DatabaseRole { get; } => PartitionKey.DatabaseRole;
```

amanda-tarafa commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.DatabaseStatistics.cs:
```diff
@@ -29,6 +29,16 @@ public sealed class DatabaseStatistics
             public DatabaseName DatabaseName { get; }
 
+            /// <summary>
+            /// Database Role of sessions this set of statistics corresponds to. Can be null.
```
```suggestion
            /// Database Role of sessions this set of statistics corresponds to. May be null, in which case the database role is ....
```

amanda-tarafa commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.DatabaseStatistics.cs:
```diff
@@ -91,9 +101,11 @@ public sealed class DatabaseStatistics
                 long readWriteTransactionRequestsPrewarmed,
                 bool healthy,
-                bool shutdown)
+                bool shutdown
+                )
```
```suggestion
                bool shutdown)
```

amanda-tarafa commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.DatabaseStatistics.cs:
```diff
@@ -91,9 +101,11 @@ public sealed class DatabaseStatistics
+                )
             {
-                DatabaseName = databaseName;
+                DatabaseName = partitionKey.DatabaseName;
+                DatabaseRole = partitionKey.DatabaseRole;
```
Better return them directly as I suggested.

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnectionStringBuilder.cs:
```diff
@@ -53,10 +53,20 @@ public sealed class SpannerConnectionStringBuilder : DbConnectionStringBuilder
         private DatabaseName _databaseName;
 
+        /// <summary>
+        /// The database role which will create the <see cref="Session"/>.
```
```suggestion
        /// The database role for the sessions created for this connection.
```

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/V1/SessionPoolTests.TargetedSessionPoolTests.cs:
```diff
@@ -76,7 +76,7 @@ private static TargetedSessionPool CreatePool(bool acquireSessionsImmediately)
                 };
                 var parent = new SessionPool(client, options);
-                return new TargetedSessionPool(parent, s_databaseName, acquireSessionsImmediately);
+                return new TargetedSessionPool(parent, new PartitionKey( s_databaseName, null), acquireSessionsImmediately);
```
We should have a test with a set database role so that we confirm that's being propagated to the acquire session request.

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/V1/SessionPoolTests.StatisticsTests.cs:
```diff
@@ -26,7 +26,8 @@ public sealed class StatisticsTests
                 var db1Stats = new DatabaseStatistics(
-                    databaseName: new DatabaseName("project", "instance", "db1"),
+                    new PartitionKey(
+                    new DatabaseName("project", "instance", "db1"), null),
```
Same here, please check that we get correct statistics if there's a database role set.

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/SpannerDatabaseRoleFixture.cs:
```diff
@@ -0,0 +1,49 @@
+        }
+    }
+
+    public SpannerConnection GetConnection(SpannerConnectionStringBuilder builder) =>
```
Where is this being used?

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/SpannerGranularAccessTest.cs:
```diff
@@ -0,0 +1,63 @@
+        string databaseRole = "reader";
+
+        // Create a database role named 'reader' that just have 'SELECT' (Read) privilege on table.
+        using var connection = _fixture.GetConnection();
+        string databaseRoleCreate = $"CREATE ROLE {databaseRole}";
+        string databaseRoleGrant = $"GRANT SELECT ON TABLE {_fixture.TableName} TO ROLE {databaseRole}";
+        var createCmd = connection.CreateDdlCommand(databaseRoleCreate, databaseRoleGrant);
+        await createCmd.ExecuteNonQueryAsync();
```
I think you can do all of this in the fixture, and just once.

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/SpannerGranularAccessTest.cs:
```diff
@@ -0,0 +1,63 @@
+        await createCmd.ExecuteNonQueryAsync();
+
+        // Try to read data from the given table using reader role.
+        using var connectionWithDbRole = _fixture.GetConnection(new SpannerConnectionStringBuilder { DatabaseRole = databaseRole, DatabaseName = _fixture.DatabaseName });
```
And then you can change this methods for several, like, GetReadOnlyConnection etc.

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/SpannerGranularAccessTest.cs:
```diff
@@ -0,0 +1,63 @@
+        _fixture = fixture;
+
+    [SkippableFact]
+    public async Task ReadWriteWithDbRoleTest()
```
Please split this test in two, one for testing that read is successfull and one for testing that write is not.

hemanshv commented:
Hey @amanda-tarafa ,
Thanks for the feedback, will work on the comments 
and update the PR.
Thanks,

hemanshv commented:
Hey @amanda-tarafa I have addressed the review comments.
Please have a look and provide your feedback.
Thanks,

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/V1/SessionPoolTests.StatisticsTests.cs:
```diff
@@ -58,14 +59,26 @@ public void AllProperties()
+                var allStats = new[] { db1Stats, db2Stats, db3Stats, db4Stats };
                 var statistics = new Statistics(allStats);
 
                 Assert.Equal(allStats, statistics.PerDatabaseStatistics);
```
I just realized, we need to deprecate `SessionPools.Statistics.PerDatabaseStatistics` and add `SessionsPools.Statistics.PerSessionPoolKeyStatistics`.

Implementation wouldn't change, is the name of the property that is wrong now.

amanda-tarafa commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.DatabaseStatistics.cs:
```diff
@@ -20,14 +20,25 @@ namespace Google.Cloud.Spanner.V1
-        /// A snapshot of statistics for one database within a <see cref="SessionPool"/>.
+        /// A snapshot of statistics for a SessionPoolKey within a <see cref="SessionPool"/>.
         /// </summary>
         public sealed class DatabaseStatistics
```
And ideally we also deprecate this class and create one exactly like this one but named `SessionPoolKeyStatistics`.

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/SpannerReaderRoleFixture.cs:
```diff
@@ -0,0 +1,68 @@
+[CollectionDefinition(nameof(SpannerReaderRoleFixture))]
+public class SpannerReaderRoleFixture : SpannerTableFixture, ICollectionFixture<SpannerReaderRoleFixture>
+{
+    public readonly string DatabaseRole = "reader";
```
```suggestion
    public const string ReaderDatabaseRole = "reader";
```

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/SpannerReaderRoleFixture.cs:
```diff
@@ -0,0 +1,68 @@
+                {"StringValue", SpannerDbType.String, "value" }
+            });
+            insertCmd.ExecuteNonQuery();
+            if (!RunningOnEmulator)
+            {
+                CreateDatabaseRoleAndGrantAccess();
+            }
```
I think I would move this to the CreateTable method.

hemanshv commented:
Hey @amanda-tarafa , 
I have addressed all the review comment apart from the ones related to stats deprecation in commit **2f51daf**.
Will address the changes related to deprecation in the next commit. 
Please have a look and provide your feedback.

Thanks.

hemanshv commented:
Yes @amanda-tarafa  thats fine
Thanks

hemanshv commented:
Hey @amanda-tarafa,

I have updated the PR with both the commits.
Please have a look and provide your feedback.

amanda-tarafa commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.Statistics.cs:
```diff
@@ -25,37 +26,45 @@ public partial class SessionPool
+            /// The statistics broken down by SessionPoolKey. This is never null, but may be empty.
             /// </summary>
-            public IReadOnlyList<DatabaseStatistics> PerDatabaseStatistics { get; }
+            [Obsolete("Use PerSessionPoolKeyStatistics instead.", error: false)]
```
```suggestion
            [Obsolete("Use PerSessionPoolKeyStatistics instead. Both properties return the same data, but PerSessionPoolKeyStatistics name better represents the fact that sessions are pooled based on aspects other than database name.")]
```

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/docs/connection_string.md:
```diff
@@ -175,4 +175,14 @@ The valid type mappings for `SpannerDbType.Date` are:
+
+## DatabaseRole
+
+The database role used while creating sessions for a database. This is only used when fine grained
+access control (FGAC) is enabled for the spanner instance otherwise this is ignored. In-case when FGAC is enabled
+and database role is not provided explicitly or may be null in that case default (public) role will be used.
```
(And remove all other references to FGAC if you have any, the acronym does not appear on public docs.)

```suggestion
The database role associated to this connection string. This option is only used if Fine Grained
Access Control is enabled for the Spanner instance. If Fine Grained Access Control is enabled for this Spanner instance, and a database role is not provided the default (public) role will be used.
```

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/docs/connection_string.md:
```diff
@@ -175,4 +175,14 @@ The valid type mappings for `SpannerDbType.Date` are:
+access control (FGAC) is enabled for the spanner instance otherwise this is ignored. In-case when FGAC is enabled
+and database role is not provided explicitly or may be null in that case default (public) role will be used.
+
+Example:
```
```suggestion
Example (given that you have created the "read" role):
```

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnection.cs:
```diff
@@ -645,22 +646,36 @@ public async Task ShutdownSessionPoolAsync(CancellationToken cancellationToken =
         /// </summary>
         /// <returns>The session pool statistics, or <c>null</c> if there is no current session pool
         /// for the database specified in the connection string.</returns>
+        [Obsolete("Use GetSessionPoolStatistics instead. SessionPool no longer maintains database specific pool.", error: false)]
```
```suggestion
        [Obsolete("Use GetSessionPoolStatistics instead. Both methods return the same data, but the GetSessionPoolStatistics name better reflects the fact that sessions are pooled on aspects other than database name.")]

```

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnection.cs:
```diff
@@ -645,22 +646,36 @@ public async Task ShutdownSessionPoolAsync(CancellationToken cancellationToken =
         }
 
+        /// <summary>
+        /// Retrieves SessionPoolKey-specific statistics for the session pool associated with the connection string. The connection string must
```
```suggestion
        /// Retrieves statistics for the session pool associated with the connection string. The connection string must
```

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnection.cs:
```diff
@@ -645,22 +646,36 @@ public async Task ShutdownSessionPoolAsync(CancellationToken cancellationToken =
 
+        /// <summary>
+        /// Retrieves SessionPoolKey-specific statistics for the session pool associated with the connection string. The connection string must
+        /// include a database name. We can also optionally provide a database role.
```
```suggestion
        /// include a database name.
```

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnection.cs:
```diff
@@ -645,22 +646,36 @@ public async Task ShutdownSessionPoolAsync(CancellationToken cancellationToken =
+        /// include a database name. We can also optionally provide a database role.
+        /// </summary>
+        /// <returns>The session pool statistics, or <c>null</c> if there is no current session pool
+        /// for the database specified in the connection string.</returns>
```
```suggestion
        /// associated to the specified connection string.</returns>

```

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SessionPoolManager.cs:
```diff
@@ -161,12 +162,20 @@ internal void Release(SessionPool sessionPool)
-            GaxPreconditions.CheckNotNull(databaseName, nameof(databaseName));
             _targetedPools.TryGetValue(options, out var targetedPool);
-            return targetedPool?.SessionPoolOrNull?.GetStatisticsSnapshot(databaseName);
+            return targetedPool?.SessionPoolOrNull?.GetStatisticsSnapshot(GaxPreconditions.CheckNotNull(key, nameof(key)));
```
Consider moving the key null validation to its own line as it was before for the database name, here and everywhere, it's hardish to read now.

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SessionPoolManager.cs:
```diff
@@ -161,12 +162,20 @@ internal void Release(SessionPool sessionPool)
+        }
+
+        [Obsolete("", error: false)]
+        internal SessionPool.DatabaseStatistics GetDatabaseStatistics(SpannerClientCreationOptions options, DatabaseName databaseName)
```
This method is internal, it can be removed and callers can now call the new one.

hemanshv replied:
This cannot be removed for now, as it is being used in SpannerConnection.GetSessionPoolDatabaseStatistics(). If we remove  this then there is no way to get DatabaseStatistics from SessionPoolManger. Although we can get SessionPoolKeyStatistics from its new overload, but then will not be able to converted it to DatabaseStatistics. Atleast not through DatabaseStatistics  contructor as its internal. Let me know your thoughts on the same.

amanda-tarafa replied:
Ah yes, you are right. I had forgotten that the Statistics classes are in the other package. But then please, even though it's internal, leave a reason for it being obsolete. Something like, "we still need to call this method for implementing the obsolete SpannerConnection.GetSessionPoolDatabaseStatistics()".

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/V1/SessionPoolTests.cs:
```diff
@@ -250,6 +252,39 @@ public async Task GetStatisticsSnapshot_MultipleDatabases()
         }
 
+        [Fact(Timeout = TestTimeoutMilliseconds)]
+        public async Task GetSessionPoolKeyStatisticsSnapshot_MultipleDatabases()
```
How is this test different from the one in line 223?

I think we might need want that has multiple database/roles pairs instead, and maybe for that you can modify the one in line 223.

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/V1/SessionPoolTests.TargetedSessionPoolTests.cs:
```diff
@@ -15,8 +15,10 @@
 using Google.Api.Gax.Grpc;
 using Google.Cloud.ClientTesting;
 using Google.Cloud.Spanner.Common.V1;
+using Google.Cloud.Spanner.V1.Internal.Logging;
```
Why do we need this here?

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/V1/SessionPoolTests.TargetedSessionPoolTests.cs:
```diff
@@ -99,7 +99,7 @@ public async Task Construction_AcquireSessionsImmediately()
                 var pool = CreatePool(true);
                 var client = (SessionTestingSpannerClient) pool.Client;
-
+          
```
Ping on this.

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/V1/SessionPoolTests.TargetedSessionPoolTests.cs:
```diff
@@ -928,6 +930,21 @@ public async Task ShutdownPoolAsync_CannotAcquireAfterShutdown()
+            [Fact]
+            public void BatchCreateSession_WithDatabaseRole()
+            {
+                Mock<SpannerClient> spannerClientMock = SpannerClientHelpers.
```
Instead of using a mock here, can't you use the same (with maybe some slight modifications) as we are using for the rest of all other tests?

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/V1/SessionPoolTests.StatisticsTests.cs:
```diff
@@ -22,11 +24,13 @@ public partial class SessionPoolTests
         {
-            [Fact]
+            [Fact]
+            [Obsolete]
```
We don't mark tests as obsolete, we use the the ignore warning directive on the piece of code that uses the deprecated code.

And also not even sure you need to mark this one as you have changed all the code to use the new class.

But, it'd still be good to maybe continue to tests the old class until we have removed it.

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Snippets/SpannerConnectionSnippets.cs:
```diff
@@ -424,8 +424,9 @@ public async Task WhenSessionPoolReady()
-        [Fact]
-        public void GetSessionPoolDatabaseStatistics()
+        [Fact]
+        [Obsolete]
```
Same comment as before re: not marking a test as Obsolete but using "ignore warning" instead.

But also, we should test the new method as well.

amanda-tarafa reviewed (changes requested):
Still a few change requests, but looking good over all. Thanks!

hemanshv commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.cs:
```diff
@@ -129,18 +129,26 @@ private void MaintainPool()
-            GaxPreconditions.CheckNotNull(databaseName, nameof(databaseName));
-            return _targetedPools.TryGetValue(databaseName, out var pool) ? pool.GetStatisticsSnapshot() : null;
+        [Obsolete("Use its overloaded method with SessionPoolKey instead.", error: false)]
+        public DatabaseStatistics GetStatisticsSnapshot(DatabaseName databaseName)
```
I think we should not deprecate methods like AcquireSessionAsync, WhenPoolReady, As they are still useful for users who are not using FGAC. They don't need to worry about what the SessionPoolKey is and all of those things. What's your thoughts on the same.

amanda-tarafa replied:
Yes, I think I agree, what I would like is to be able to have a `SessionPoolKeyStatistics GetStatisticsSnapshot(DatabaseName name)` which we cannot because we have the obsolete one.

Maybe the new ones should be called `GetPoolStatisticsSnapshot` and then we could have one receiving the `SessionPoolKey` and another one receiving just the `DatabaseName` and both returning `SessionPoolKeyStatistics`.

amanda-tarafa commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.cs:
```diff
@@ -129,18 +129,26 @@ private void MaintainPool()
+            new Statistics(_targetedPools.ToArray().Select(p => p.Value.GetStatisticsSnapshot()).ToList().AsReadOnly());
+
         /// <summary>
         /// Provides a snapshot of statistics for a database-specific pool.
```
Can you please add here (and in all the overloads that receive just the DatabaseName) something like:
```suggestion
        /// Provides a snapshot of statistics for a database-specific pool.
        /// This is equivalent to calling <see cref="GetStatisticsSnapshot(SessionPoolKey)"> passing a session pool key
        /// with a null database role.
```

hemanshv commented:
Hey @amanda-tarafa ,
I have updated the PR, please have a look and  provide your feedback.
Thanks,

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Snippets/SpannerConnectionSnippets.cs:
```diff
@@ -424,16 +424,46 @@ public async Task WhenSessionPoolReady()
+            // Additional: GetSessionPoolDatabaseStatistics
+            using (SpannerConnection connection = new SpannerConnection(connectionString))
+            {
+#pragma warning disable CS0618 // Type or member is obsolete
```
Let's update the samples to use the non-deprecated methods instead.

And please check the docs where these samples are included in case they need updating too. Ping me later to tell you how to find the docs where these samples are being shown.

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Snippets/SpannerConnectionSnippets.cs:
```diff
@@ -424,16 +424,46 @@ public async Task WhenSessionPoolReady()
         {
             string connectionString = _fixture.ConnectionString;
 
             // Sample: GetSessionPoolDatabaseStatistics
```
We might need/want to change these names. Check with me later to point you in the right direction.

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/V1/SessionPoolTests.cs:
```diff
@@ -63,8 +64,10 @@ public async Task SessionRefreshOnRelease()
-                Logger.DefaultLogger.Info(pool.GetStatisticsSnapshot(_fixture.DatabaseName).ToString());
+                await pool.WhenPoolReady(_fixture.DatabaseName, cts.Token);
+#pragma warning disable CS0618 // Type or member is obsolete
+                Logger.DefaultLogger.Info(pool.GetStatisticsSnapshot(_fixture.DatabaseName).ToString());
```
Use the non deprecated version instead.

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/SpannerStressTests.cs:
```diff
@@ -158,12 +159,14 @@ private async Task RunStress(Func<SpannerConnectionStringBuilder, Task> func)
+            await pool.WhenPoolReady(_fixture.DatabaseName, cancellationToken);
+
+#pragma warning disable CS0618 // Type or member is obsolete
+            logger.Info($"Prewarm complete. Pool stats: {pool.GetStatisticsSnapshot(_fixture.DatabaseName)}");
```
Use the non deprecated version instead.

amanda-tarafa reviewed (changes requested):
A few changes around the changes you last made.

Also, a couple of changes were left over from the last review, make sure you cover all requests before pushing.

amanda-tarafa commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.cs:
```diff
@@ -129,30 +129,65 @@ private void MaintainPool()
+        }
+
+        /// <summary>
+        /// Provides a snapshot of statistics for pool associated with the given <see cref="SessionPoolKey"/>.
```
Change here and everywhere.

```suggestion
        /// Provides a snapshot of statistics for the pool associated with the given <see cref="SessionPoolKey"/>.
```

amanda-tarafa commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.cs:
```diff
@@ -129,30 +129,65 @@ private void MaintainPool()
+        }
+
+        /// <summary>
+        /// Provides a snapshot of statistics for pool associated with the given database.
```
```suggestion
        /// Provides a snapshot of statistics for the pool associated with the given database.
```

amanda-tarafa commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.SessionPoolKey.cs:
```diff
@@ -0,0 +1,88 @@
+        public string DatabaseRole { get; }
+
+        /// <summary>
+        ///  Instantiates a new instance of the <see cref="SessionPoolKey"/> object.
```
```suggestion
        ///  Creates a new instance of the <see cref="SessionPoolKey"/> object.
```

jskeet replied:
Either change "object" to "class", or just `Creates a new <see cref="SessionPoolKey" />.`

amanda-tarafa commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.SessionPoolKey.cs:
```diff
@@ -0,0 +1,88 @@
+        /// <summary>
+        /// Creates a <see cref="SessionPoolKey"/> with the given database name. 
+        /// </summary>
+        /// <param name="databaseName">The database name for the associated pool. Must not be null</param>
```
```suggestion
        /// <param name="databaseName">The database name for the associated pool. Must not be null.</param>

```

amanda-tarafa commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.SessionPoolKey.cs:
```diff
@@ -0,0 +1,88 @@
+        /// Creates a <see cref="SessionPoolKey"/> with the given database name. 
+        /// </summary>
+        /// <param name="databaseName">The database name for the associated pool. Must not be null</param>
+        /// <returns></returns>
```
If empty, remove it. Check here and everywhere.

amanda-tarafa commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.SessionPoolKey.cs:
```diff
@@ -0,0 +1,88 @@
+            new SessionPoolKey(databaseName, null);
+
+        /// <summary>
+        /// Creates a <see cref="SessionPoolKey"/> with the given database role. 
```
```suggestion
        /// Creates a <see cref="SessionPoolKey"/> based on this one but with the given database role. 

```

amanda-tarafa commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.DatabaseStatistics.cs:
```diff
@@ -14,95 +14,87 @@
+            public DatabaseName DatabaseName => PoolKeyStatistics .DatabaseName;
+
+            /// <summary>
+            /// Database Role of sessions this set of statistics corresponds to.  May be null
```
```suggestion
            /// The database role of the sessions in the session pool this set of statistics corresponds to.  May be null
```

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/docs/index.md:
```diff
@@ -251,7 +251,7 @@ provides a simple method to accomplish this:
 If you want to diagnose session management issues,
 `SpannerConnection` provides a method to fetch a snapshot of
-the statistics for the session pool associated with the database
+the statistics for the session pool associated with the SessionPoolKey
 associated with the connection:
```
```suggestion
the statistics for the session pool associated with the connection:
```

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Snippets/SpannerConnectionSnippets.cs:
```diff
@@ -425,16 +425,44 @@ public async Task WhenSessionPoolReady()
         {
             string connectionString = _fixture.ConnectionString;
 
             // Sample: GetSessionPoolDatabaseStatistics
             // Additional: GetSessionPoolDatabaseStatistics
```
You need to change these, else the docs won't pick it up properly.

```suggestion
            // Sample: GetSessionPoolStatistics
            // Additional: GetSessionPoolStatistics
```

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/docs/index.md:
```diff
@@ -251,7 +251,7 @@ provides a simple method to accomplish this:
 associated with the connection:
 
-{{sample:SpannerConnection.GetSessionPoolDatabaseStatistics}}
+{{sample:SpannerConnection.GetSessionPoolKeyStatistics}}
```
```suggestion
{{sample:SpannerConnection.GetSessionPoolStatistics}}
```

jskeet replied:
Additionally, it's worth building the docs to check this:

- From the top level:
- `./build.sh Google.Cloud.Spanner.Data`
- `cd docs`
- `./builddocs.sh Google.Cloud.Spanner.Data`
- `../packages/docfx.2.59.4/docfx serve output/Google.Cloud.Spanner.Data/site`
- Open `localhost:8080` and check the docs look okay

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Snippets/SpannerConnectionSnippets.cs:
```diff
@@ -425,16 +425,44 @@ public async Task WhenSessionPoolReady()
+        }
+
+        [Fact]
+        public void GetSessionPoolKeyStatistics()
```
We don't need two samples that do the same. Keep the one above with my suggestion on sample name, etc.

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnection.cs:
```diff
@@ -647,7 +648,7 @@ public async Task WhenSessionPoolReady(CancellationToken cancellationToken = def
             DatabaseName databaseName = Builder.DatabaseName;
             GaxPreconditions.CheckState(databaseName != null, $"{nameof(WhenSessionPoolReady)} cannot be used without a database.");
```
This check is now on GetSessionKey, you can remove it from here, and from all the rest.
You also don't need to store the databaseName in a variable now, as it will only be used on GetSessionKey.
And for readability, and for avoiding opening the conneciton unnecessariyl obtain the key first.

Do this for all other methods as well.

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/V1/SessionPoolTests.cs:
```diff
@@ -207,44 +214,45 @@ public async Task WhenPoolReady()
+            var sessionPool = new SessionPool(client, options);
+            // We haven't used the database in this session pool, so there are no statistics for it.
+#pragma warning disable CS0618 // Type or member is obsolete
+            Assert.Null(sessionPool.GetStatisticsSnapshot(s_sampleDatabaseName));
```
Use the method version that is not deprecated instead.

amanda-tarafa reviewed (changes requested):
These are mostly types and documentation improvements.

@jskeet feel free to take a look now.

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/V1/SessionPoolTests.cs:
```diff
@@ -23,7 +23,8 @@
 using System.Threading.Tasks;
 using Xunit;
-
+using static Google.Cloud.Spanner.V1.SessionPool;
```
Why do we need this additional using directive?

jskeet replied:
Ping for this.

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/V1/SessionPoolTests.TargetedSessionPoolTests.cs:
```diff
@@ -928,6 +929,21 @@ public async Task ShutdownPoolAsync_CannotAcquireAfterShutdown()
+
+                BatchCreateSessionsRequest batchCreateSessionsRequest;
+                Assert.True(spannerClientMock.BatchCreateSessionRequests.TryDequeue(out batchCreateSessionsRequest));
+                Assert.Equal(s_databaseRole, batchCreateSessionsRequest.SessionTemplate.CreatorRole);
```
Assert the database name as well, just to make sure that `SessionPoolKey.WithDatabaseRole` doesn't lose that?

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/V1/SessionPoolTests.cs:
```diff
@@ -35,6 +36,7 @@ public partial class SessionPoolTests
 
         private static readonly DatabaseName s_sampleDatabaseName = new DatabaseName("project", "instance", "database");
         private static readonly DatabaseName s_sampleDatabaseName2 = new DatabaseName("project", "instance", "database2");
+        private static readonly string s_DatabaseRole = "testrole";
```
Please rename to s_databaseRole

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/V1/SessionPoolTests.cs:
```diff
@@ -226,25 +243,61 @@ public async Task GetStatisticsSnapshot_MultipleDatabases()
+        }
+
+        [Fact(Timeout = TestTimeoutMilliseconds)]
+        public async Task GetPoolStatisticsSnapshot_MultipleDatabases()
```
Not sure which test you're suggesting we get rid of. (Line 238 is in the middle of a test.)

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SessionPoolManager.cs:
```diff
@@ -242,9 +252,9 @@ public override string ToString()
                 else
                 {
-                    foreach (var databaseStatistics in SessionPoolStatistics.PerDatabaseStatistics)
+                    foreach (var sessionpoolStatistics in SessionPoolStatistics.PerSessionPoolKeyStatistics)
```
Nit: rename sessionpoolStatistics to sessionPoolStatistics (capital P)

jskeet commented on apis/Google.Cloud.Spanner.Data/docs/connection_string.md:
```diff
@@ -175,4 +175,14 @@ The valid type mappings for `SpannerDbType.Date` are:
+
+Example (given you have created the "read" role):
+
+- DatabaseRole="read"
```
Wouldn't it just be `DatabaseRole=read` (without the double quotes)?

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.DatabaseStatistics.cs:
```diff
@@ -14,95 +14,87 @@
             /// The number of read-only sessions in the pool.
             /// </summary>
-            public int ReadPoolCount { get; }
+            public int ReadPoolCount => PoolKeyStatistics .ReadPoolCount;
```
Please remove the space before the period in all of these expressions.

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.SessionPoolKey.cs:
```diff
@@ -0,0 +1,88 @@
+
+        /// <inheritdoc/>
+        public override int GetHashCode() =>
+            DatabaseRole is null ? DatabaseName.GetHashCode() :
```
This can be a single line with the null conditional operator:

```csharp
GaxEqualityHelpers.CombineHashCodes(DatabaseName.GetHashCode(), DatabaseRole?.GetHashCode());
```

Note that I've reordered them so that we use name before role everywhere.

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.SessionPoolKey.cs:
```diff
@@ -0,0 +1,88 @@
+        /// </summary>
+        /// <returns><c>true</c> if the specified object is equal to this instance; otherwise <c>false</c>.</returns>
+        public bool Equals(SessionPoolKey other) =>
+            (other is null) ? false : DatabaseName.Equals(other.DatabaseName) && DatabaseRole == other.DatabaseRole;
```
Change the `.Equals` to `==` for simplicity.

amanda-tarafa replied:
Oh, I asked for this to be Equals, do resource names redefine the equality operators? Note this is not the string, but the resource name?

jskeet replied:
I hadn't noticed that it's a DatabaseName. But yes, we *do* overload equality operators. See https://cloud.google.com/dotnet/docs/reference/Google.Cloud.Spanner.Common.V1/latest/Google.Cloud.Spanner.Common.V1.DatabaseName#Google_Cloud_Spanner_Common_V1_DatabaseName_op_Equality_Google_Cloud_Spanner_Common_V1_DatabaseName_Google_Cloud_Spanner_Common_V1_DatabaseName_

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.SessionPoolKeyStatistics.cs:
```diff
@@ -0,0 +1,126 @@
+    /// <summary>
+    /// A snapshot of statistics for a <see cref="SessionPoolKey"/> within a <see cref="SessionPool"/>.
+    /// </summary>
+    public sealed class SessionPoolKeyStatistics
```
This makes it sound like it's statistics about the key itself. I would suggest SessionPoolPartitionStatistics, but as Amanda has said, partition has a particular meaning in Spanner.

If we can find an appropriate name, we might also want to rename SessionPoolKey similarly, to link the two by name - e.g. "SessionPoolSubdivisionKey" and "SessionPoolSubdivisionStatistics". (That's just an example, not a concrete suggestion.)

We should brainstorm the name further.

jskeet replied:
(If we rename SessionPoolKey that has a bunch of follow-on changes, of course.)

jskeet replied:
SessionPoolSegmentStatistics / SessionPoolSegmentKey would work.

hemanshv replied:
Hey @jskeet & @amanda-tarafa ,
Will it be fine to use SessionPoolSegmentStatistics / SessionPoolSegmentKey.

jskeet replied:
I've just checked with Amanda - we're both happy with that. Apologies for the extra work :( I'm happy to re-review once you've renamed, to check that we haven't missed anything. (It'll be very easy to miss names *based* on the previous ones.)

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.SessionPoolKeyStatistics.cs:
```diff
@@ -0,0 +1,126 @@
+    public sealed class SessionPoolKeyStatistics
+    {
+        /// <summary>
+        /// The database this set of statistics corresponds to.
```
I'd go with "associated with" rather than "associated to" but yes.

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.SessionPoolKeyStatistics.cs:
```diff
@@ -0,0 +1,126 @@
+    /// </summary>
+    public sealed class SessionPoolKeyStatistics
+    {
+        /// <summary>
```
If we're exposing the key publicly, do we need these properties at all? Users can just get to them via the key.

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.cs:
```diff
@@ -129,30 +129,65 @@ private void MaintainPool()
+        /// Provides a snapshot of statistics for pool associated with the given <see cref="SessionPoolKey"/>.
+        /// </summary>
+        /// <returns>A snapshot of statistics for this pool.</returns>
+        public SessionPoolKeyStatistics GetPoolStatisticsSnapshot(SessionPoolKey key)
```
Once we have a good name for the "partition" concept, we might want to adjust this name. `GetSegmentStatisticsSnapshot` for example. (We're already in SessionPool, so the pool-ness seems clear already. Emphasizing that it's a segment of the pool feels useful.)

jskeet reviewed (commented):
No behavioral suggestions - but the naming part is important.

We might want to have a VC with any interested folks about the naming - I've really warmed to "segment" as something we could talk about in docs and use consistently. (I'm not suggesting renaming TargetedSessionPool and other internals yet, but we *could* do so.)

jskeet commented:
Can this PR now be marked "ready for review" if we think it's ready to merge on successful review?

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/V1/SessionPoolTests.TargetedSessionPoolTests.cs:
```diff
@@ -76,7 +77,7 @@ private static TargetedSessionPool CreatePool(bool acquireSessionsImmediately)
                 };
                 var parent = new SessionPool(client, options);
-                return new TargetedSessionPool(parent, s_databaseName, acquireSessionsImmediately);
+                return new TargetedSessionPool(parent, SessionPoolSegmentKey.Create( s_databaseName), acquireSessionsImmediately);
```
Nit: remove space after `Create(`

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/V1/SessionPoolTests.cs:
```diff
@@ -24,6 +24,7 @@
 using System.Linq;
 using System.Threading.Tasks;
 using Xunit;
+using static Google.Cloud.Spanner.V1.SessionPool;
```
I'm not sure why this is needed. If it is, that's fine - but worth checking.

hemanshv replied:
@jskeet  Its required because we are creating a SessionPoolSegementKey() on line 236. If we omit this we need to create it like SessionPool.SessionPoolSegmentKey().

jskeet replied:
Ah right. Yes, that makes sense. (I'm still unclear why the on in integration tests is needed though.)

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/V1/SessionPoolTests.cs:
```diff
@@ -207,44 +214,43 @@ public async Task WhenPoolReady()
-        [Fact(Timeout = TestTimeoutMilliseconds)]
-        public async Task GetStatisticsSnapshot_MultipleDatabases()
+        [Fact(Timeout = TestTimeoutMilliseconds)]
+        public async Task GetPoolStatisticsSnapshot_MultipleDatabases()
```
Undo this rename? We're still calling GetStatisticsSnapshot.

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/V1/SessionPoolTests.cs:
```diff
@@ -207,44 +214,43 @@ public async Task WhenPoolReady()
         }
 
-        [Fact]
+        [Fact]
```
This test should be renamed to reflect the method name change.

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SessionPoolManager.cs:
```diff
@@ -242,9 +252,9 @@ public override string ToString()
                 else
                 {
-                    foreach (var databaseStatistics in SessionPoolStatistics.PerDatabaseStatistics)
+                    foreach (var sessionPoolStatistics in SessionPoolStatistics.PerSessionPoolSegmentStatistics)
```
Completely optional: could rename this variable to `segmentStatistics`

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnection.cs:
```diff
@@ -661,25 +661,44 @@ public async Task WhenSessionPoolReady(CancellationToken cancellationToken = def
+        /// Retrieves statistics for the session pool associated with the connection string. The connection string must
+        /// include a database name.
+        /// </summary>
+        /// <returns>The session pool statistics, or <c>null</c> if there is no current session pool
```
We should clarify that it's the session pool statistics associated with the corresponding segment - not the whole session pool.
Indeed, it should possibly be called GetSessionPoolSegmentStatistics, which would match the return type.

@amanda-tarafa what do you think?

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.DatabaseStatistics.cs:
```diff
@@ -14,95 +14,87 @@
+            /// </summary>
+            public string DatabaseRole => PoolSegmentStatistics.DatabaseRole;
+
+            private SessionPoolSegmentStatistics PoolSegmentStatistics  { get; }
```
I'd just call this SegmentStatistics. ("Session pool" feels like a single concept here which we shouldn't just have half of. I think it's okay for that to be implicit.)

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.DatabaseStatistics.cs:
```diff
@@ -14,95 +14,87 @@
-                ReadWriteTransactionRequestsPrewarmed = readWriteTransactionRequestsPrewarmed;
-                Healthy = healthy;
-                Shutdown = shutdown;
+                PoolSegmentStatistics = sessionPoolSegmentStatistics;
```
(Very optional) This constructor could now be expression-bodied:

```csharp
internal DatabaseStatistics(SessionPoolSegmentStatistics sessionPoolSegmentStatistics) =>
    SegmentStatistics = sessionPoolSegmentStatistics;
```

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.DatabaseStatistics.cs:
```diff
@@ -111,7 +103,7 @@ public override string ToString()
                     : Healthy ? "Healthy"
                     : "Unhealthy";
-                return $"Database: {DatabaseName}; Active: {ActiveSessionCount}; Read pool: {ReadPoolCount}; Write pool: {ReadWritePoolCount}; In-flight creation: {InFlightCreationCount}; Pending acquisitions: {PendingAcquisitionCount}; Status: {status}; Read/write prewarming: {ReadWriteTransactionRequestsPrewarmed}/{ReadWriteTransactionRequests}";
+                return $"{PoolSegmentStatistics}; Active: {ActiveSessionCount}; Read pool: {ReadPoolCount}; Write pool: {ReadWritePoolCount}; In-flight creation: {InFlightCreationCount}; Pending acquisitions: {PendingAcquisitionCount}; Status: {status}; Read/write prewarming: {ReadWriteTransactionRequestsPrewarmed}/{ReadWriteTransactionRequests}";
```
The "{PoolSegmentStatistics" part is going to give the same information twice.

Can't we just have `public override string ToString() => PoolSegmentStatistics.ToString();`?

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.SessionPoolSegmentKey.cs:
```diff
@@ -0,0 +1,85 @@
+public sealed partial class SessionPool
+{
+    /// <summary>
+    /// Aspects under which the sessions are pooled.
```
```suggestion
    /// Aspects under which a session pool is segmented.
```

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.SessionPoolSegmentKey.cs:
```diff
@@ -0,0 +1,85 @@
+    /// <summary>
+    /// Aspects under which the sessions are pooled.
+    /// </summary>
+    public class SessionPoolSegmentKey : IEquatable<SessionPoolSegmentKey>
```
We should make this class sealed

amanda-tarafa replied:
+1

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.Statistics.cs:
```diff
@@ -25,37 +26,46 @@ public partial class SessionPool
+            /// The statistics broken down by <see cref="SessionPoolSegmentKey"/>. This is never null, but may be empty.
             /// </summary>
-            public IReadOnlyList<DatabaseStatistics> PerDatabaseStatistics { get; }
+            [Obsolete("Use PerSessionPoolKeyStatistics instead. Both properties return the same data, but PerSessionPoolKeyStatistics name" +
```
Modify comment to use PerSessionPoolSegmentStatistics (or PerSegmentStatistics)

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.Statistics.cs:
```diff
@@ -25,37 +26,46 @@ public partial class SessionPool
+            /// <summary>
+            /// The statistics broken down by database. This is never null, but may be empty.
+            ///</summary>
+            public IReadOnlyList<SessionPoolSegmentStatistics> PerSessionPoolSegmentStatistics { get; }
```
I think this could just be PerSegmentStatistics (corresponding to the previous PerDatabaseStatistics). We're already in SessionPool.Statistics; I think the session-pool-ness is implicit.

amanda-tarafa replied:
+1

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.Statistics.cs:
```diff
@@ -25,37 +26,46 @@ public partial class SessionPool
+            ///</summary>
+            public IReadOnlyList<SessionPoolSegmentStatistics> PerSessionPoolSegmentStatistics { get; }
+
+            internal Statistics(IReadOnlyList<SessionPoolSegmentStatistics> perSessionPoolSegmentStatistics)
```
I'd make this expression-bodied.

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.Statistics.cs:
```diff
@@ -25,37 +26,46 @@ public partial class SessionPool
-            internal Statistics(IReadOnlyList<DatabaseStatistics> perDatabaseStatistics) =>
-                PerDatabaseStatistics = perDatabaseStatistics;
+            /// <summary>
+            /// The statistics broken down by database. This is never null, but may be empty.
```
Update "database" to instead refer to SessionPoolSegmentKey. (That's nicely future-proof.)

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.TargetedSessionPool.cs:
```diff
@@ -36,7 +36,7 @@ internal sealed class TargetedSessionPool : SessionPoolBase
             // Read-only state
             private readonly object _lock = new object();
-            private readonly DatabaseName _databaseName;
+            private readonly SessionPoolSegmentKey _sessionPoolSegmentKey;
```
I'd probably just call it _segmentKey, but that doesn't really matter and can be changed later - leave or change as you see fit.

(I'm probably being very inconsistent here, too!)

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.cs:
```diff
@@ -129,30 +129,65 @@ private void MaintainPool()
         public Statistics GetStatisticsSnapshot() =>
-            new Statistics(_targetedPools.ToArray().Select(p => p.Value.GetStatisticsSnapshot()).ToList().AsReadOnly());
-
+            new Statistics(_targetedPools.ToArray().Select(p => p.Value.GetStatisticsSnapshot()).ToList().AsReadOnly());
```
I can't see what's changed here. Any ideas? (Maybe line endings?)

jskeet reviewed (commented):
Only really small nits now, although there are a couple of "real" changes (renaming a method and sealing a class, IIRC).

amanda-tarafa commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.cs:
```diff
@@ -129,30 +129,65 @@ private void MaintainPool()
+
         /// <summary>
         /// Provides a snapshot of statistics for a database-specific pool.
+        /// This is equivalent to calling <see cref="GetSegmentStatisticsSnapshot(SessionPoolSegmentKey)"/> passing a session pool key
```
Check all the others as well
```suggestion
        /// This is equivalent to calling <see cref="GetSegmentStatisticsSnapshot(SessionPoolSegmentKey)"/> passing a segment key
```

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnection.cs:
```diff
@@ -644,10 +645,9 @@ private void Open(Action transactionEnlister)
-            GaxPreconditions.CheckState(databaseName != null, $"{nameof(WhenSessionPoolReady)} cannot be used without a database.");
             await OpenAsync(cancellationToken).ConfigureAwait(false);
-            await _sessionPool.WhenPoolReady(databaseName, cancellationToken).ConfigureAwait(false);
+            var sessionPoolSegmentKey = GetSessionPoolSegmentKey(nameof(WhenSessionPoolReady));
```
Move this to before the call to OpenAsync.
We don't want to attempt to opne the connection if there's no database specified. Connection opening will fail, but the error will be off.

Do this here and in all the others.

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnection.cs:
```diff
@@ -661,25 +661,44 @@ public async Task WhenSessionPoolReady(CancellationToken cancellationToken = def
+        /// include a database name.
+        /// </summary>
+        /// <returns>The session pool statistics, or <c>null</c> if there is no current session pool
+        /// associated to the <see cref="SessionPoolSegmentKey"/>.</returns>
```
Nit: associated to => associated with (as in line 684). But we can fix that later easily.

jskeet commented on apis/Google.Cloud.Spanner.Data/docs/connection_string.md:
```diff
@@ -175,4 +175,14 @@ The valid type mappings for `SpannerDbType.Date` are:
+
+## DatabaseRole
+
+The database role associated to this connection string. This option is only used if Fine Grained
```
associated to => associated with - or just "Option to configure the database role for this connection." (That would be more in-keeping with the options above.)

jskeet reviewed (approved):
I've *only* checked my comments. (It's hard to tell exactly what's changed since the last round of reviews.) Let me know if there are any other things you want me to check.

Please wait for @amanda-tarafa to review as well before merging, of course.

amanda-tarafa reviewed (changes requested):
I think docs won't pickup the `Additional` tag on the snippet.
Please run locally the command as suggested by @jskeet and confir that docs are being generated correctly.

Otherwise LGTM

hemanshv commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Snippets/SpannerConnectionSnippets.cs:
```diff
@@ -425,16 +425,16 @@ public async Task WhenSessionPoolReady()
 
-            // Sample: GetSessionPoolDatabaseStatistics
-            // Additional: GetSessionPoolDatabaseStatistics
+            // Sample: GetSessionPoolStatistics
+            // Additional: GetSessionPoolStatistics
```
Hey @amanda-tarafa,
Thanks for reviewing the PR, I already ran the docs locally and that were working as expected.
I check it again after renaming and it seems to be working fine. Will update the PR.

Thanks,

amanda-tarafa reviewed (approved):
LGTM.

@hemanshv before merging, rebase everything into a single commit please.
